### PR TITLE
docs: Prettify JSON output for default objects in docs 

### LINKED ISSUE #15787: Prettify JSON output for default objects in docs
### What would you like Renovate to be able to do?

PR #15637 added a new feature, that shows the default objects and array in our configuration docs. But the output is hard to understand sometimes.

For example, check out [Renovate docs, `lockfileMaintenance`](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance):

json-output-lockfile-maintenance

### If you have any ideas on how this should be implemented, please tell us here.

Run Prettier to transform the JSON:

**Input:**

```json
{"enabled":false,"recreateClosed":true,"rebaseStalePrs":true,"branchTopic":"lock-file-maintenance","commitMessageAction":"Lock  file  maintenance","commitMessageTopic":null,"commitMessageExtra":null,"schedule":["before  5am on monday"],"groupName":null,"prBodyDefinitions":{"Change":"All  locks refreshed"}}
```

**Output:**

```json
{
  "enabled": false,
  "recreateClosed": true,
  "rebaseStalePrs": true,
  "branchTopic": "lock-file-maintenance",
  "commitMessageAction": "Lock  file  maintenance",
  "commitMessageTopic": null,
  "commitMessageExtra": null,
  "schedule": ["before  5am on monday"],
  "groupName": null,
  "prBodyDefinitions": { "Change": "All  locks refreshed" }
}
```

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
Now we generate html tables instead of markdown tables.
This allows us to have syntax highlight with formatted obj/arrays.

## Context
Closes #15787

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]
Only used for docs generation, right?

**Comment by Gabriel-Ladzaretti** [comment]
> Only used for docs generation, right?

Yes, this only effects these auto generated files:
`configuration-options.md`  `self-hosted-configuration.md`

**Comment by Gabriel-Ladzaretti** [comment]
full markdown for [configuration-options.md
](https://github.com/ladzaretti/15906-undefined-dash/pull/3#issuecomment-1152904609
)

full markdown for [self-hosted-configuration.md](https://github.com/ladzaretti/15906-undefined-dash/pull/3#issuecomment-1152904831)

**Comment by Gabriel-Ladzaretti** [comment]
```json
[
  "at any time"
]
```
Above is what we have for objects with only one key.
Do we want to have below instead?
```json
[ "at any time" ]
```

**Comment by rarkins** [comment]
Yes, could you consider the "pretty" JSON formatter?

**Comment by rarkins** [comment]
Something like https://www.npmjs.com/package/json-stringify-pretty-compact v3

File: tools/docs/config.ts [hunk 1]
```diff
@@ -1,11 +1,28 @@
...
 import { getEnvName } from '../../lib/workers/global/config/parse/env';
 import { readFile, updateFile } from '../utils';
 
 const options = getOptions();
 
+function buildHtmlTable(data: string[][]): string {
+  const ind = '  '; // indentation 2 spaces
+  let table = `<table>\n`;
+  for (const [i, row] of data.entries()) {
+    table += `${ind}<tr>\n`;
```
Comment (astellingwerf) [inline tools/docs/config.ts:12]: I believe that the first table row should be wrapper in a `thead` element, and the rest should be contained by a `tbody` element. That's how the current documentation is rendered anyway.

**Comment by Gabriel-Ladzaretti** [comment]
Updated with latest commit:
full markdown for [`configuration-options.md`](https://github.com/ladzaretti/15906-undefined-dash/pull/3#issuecomment-1152959958)
full markdown for [`self-hosted-configuration.md`](https://github.com/ladzaretti/15906-undefined-dash/pull/3#issuecomment-1152960094)


generated table source:
```html
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>type</td>
      <td>array</td>
    </tr>
    <tr>
      <td>subType</td>
      <td>string</td>
    </tr>
    <tr>
      <td>cli</td>
      <td><code>--automerge-schedule</code></td>
    </tr>
    <tr>
      <td>default</td>
      <td><pre lang="json">["at any time"]</pre></td>
    </tr>
  </tbody>
</table>
```

**Comment by rarkins** [comment]
@Gabriel-Ladzaretti I think I'm missing something from your linked markdown. e.g. I see this without any default value:

image

**Comment by Gabriel-Ladzaretti** [comment]
@rarkins 
a bug maybe? looks unrelated to this PR

https://docs.renovatebot.com/configuration-options/#constraints
image

**Comment by rarkins** [comment]
I was just looking for an example of changes in the PR actually. I thing by your PR was to change this behaviour. Which field should I look at?

**Comment by Gabriel-Ladzaretti** [comment]
> I was just looking for an example of changes in the PR actually. I thing by your PR was to change this behaviour. Which field should I look at?

I don't quite follow, what do you mean by "field"?
from what i understand, empty/null defaults for objects and arrays are being filtered out ¯\\_(ツ)_/¯.

**Comment by rarkins** [comment]
I meant a config option name. Is there a good one to look at for an example?

**Comment by Gabriel-Ladzaretti** [comment]
> I meant a config option name. Is there a good one to look at for an example?

define good 😅, here's the complete list.

  ## configuration-options     
  `allowedPostUpgradeCommands`
       "default": []
  `commands`
       "default": []
  `composerIgnorePlatformReqs`
       "default": []
  `constraints`
       "default": {}
  `customEnvVariables`        
       "default": {}
  `defaultRegistryUrls`       
       "default": null
  `dependencyDashboardLabels` 
       "default": null
  `dotnet`
       "default": {}
  `encrypted`
       "default": null
  `excludeCommitPaths`        
       "default": []
  `fileFilters`
       "default": []
  `includePaths`
       "default": []
  `java`
       "default": {}
  `js`
       "default": {}
  `major`
       "default": {}
  `migratePresets`
       "default": {}
  `minor`
       "default": {}
  `patch`
       "default": {}
  `php`
       "default": {}
  `postUpdateOptions`
       "default": []
  `prBodyNotes`
       "default": []
  `python`
       "default": {}
  `regexManagers`
       "default": []
  `registryAliases`
       "default": {}
  `registryUrls`
       "default": null
  `ruby`
       "default": {}
  `rust`
       "default": {}
  `secrets`
       "default": {}
  ## self-hosted-configuration
  `allowedPostUpgradeCommands`
       "default": []
  `commands`
       "default": []
  `composerIgnorePlatformReqs`
       "default": []
  `constraints`
       "default": {}
  `customEnvVariables`
       "default": {}
  `defaultRegistryUrls`
       "default": null
  `dependencyDashboardLabels`
       "default": null
  `dotnet`
       "default": {}
  `encrypted`
       "default": null
  `excludeCommitPaths`
       "default": []
  `fileFilters`
       "default": []
  `includePaths`
       "default": []
  `java`
       "default": {}
  `js`
       "default": {}
  `major`
       "default": {}
  `migratePresets`
       "default": {}
  `minor`
       "default": {}
  `patch`
       "default": {}
  `php`
       "default": {}
  `postUpdateOptions`
       "default": []
  `prBodyNotes`
       "default": []
  `python`
       "default": {}
  `regexManagers`
       "default": []
  `registryAliases`
       "default": {}
  `registryUrls`
       "default": null
  `ruby`
       "default": {}
  `rust`
       "default": {}
  `secrets`
       "default": {}

File: tools/docs/config.ts [hunk 2]
```diff
@@ -1,25 +1,45 @@
...
 import { readFile, updateFile } from '../utils';
 
 const options = getOptions();
 
+function indent(amount: number, indent = '  '): string {
+  let indentation = '';
+  for (let i = 0; i < amount; i++) {
+    indentation += indent;
+  }
+  return indentation;
```
Comment (astellingwerf) [inline tools/docs/config.ts:14]: Can't we use [string.repeat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/repeat)?

Comment (Gabriel-Ladzaretti) [reply tools/docs/config.ts]: That's clean, Didn't know this one. thanks

**Comment by rarkins** [comment]
Can we have a new example markdown output to review?

**Comment by Gabriel-Ladzaretti** [comment]
Updated markdown.
Please note that we still don't print empty/null defaults for objects/arrays.

full markdown for [`configuration-options.md`](https://github.com/ladzaretti/15569-branchPrefix-deleted/pull/1#issuecomment-1153190410
)
full markdown for [`self-hosted-configuration.md`](https://github.com/ladzaretti/15569-branchPrefix-deleted/pull/1#issuecomment-1153190828)

**Review by rarkins — DISMISSED** [review]

**Comment by rarkins** [comment]
@viceice I think let's see how this looks when inlined into the table

File: tools/docs/config.ts [hunk 3]
```diff
@@ -1,11 +1,44 @@
-import table from 'markdown-table';
+import stringify from 'json-stringify-pretty-compact';
 import { getOptions } from '../../lib/config/options';
 import { getCliName } from '../../lib/workers/global/config/parse/cli';
 import { getEnvName } from '../../lib/workers/global/config/parse/env';
 import { readFile, updateFile } from '../utils';
 
 const options = getOptions();
 
+function indent(amount: number, indent = '  '): string {
```
Comment (viceice) [inline tools/docs/config.ts:9]: i think this is a good candidate for a tagged  template function 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates

File: tools/docs/config.ts [hunk 4]
```diff
@@ -51,7 +84,7 @@ function genTable(obj: [string, string][], type: string, def: any): string {
         if (Object.keys(el[1] ?? []).length === 0) {
           return;
         }
-        el[1] = `\`${JSON.stringify(el[1])}\``;
+        el[1] = `<pre lang="json">${stringify(el[1], { indent: 2 })}</pre>`;
```
Comment (viceice) [inline tools/docs/config.ts:87]: does this have syntax highlight in final docs?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: tools/docs/config.ts (hunk 4 above)
Comment (ladzaretti) [reply tools/docs/config.ts:87]: Looks like it does not :[
image

looks good though regardless imho when compared to:

image

also, we are currently also using html tables, so there must be some kind translation process from code generated markdown tables to the final html docs.
so, would generating the docs with html tables from the get go could be of a problem?

Comment (rarkins) [reply tools/docs/config.ts:87]: I still prefer the new screenshot to the old

**Comment by HonkingGoose** [comment]
> @Gabriel-Ladzaretti I think I'm missing something from your linked markdown. e.g. I see this without any default value:
> 
> image

@rarkins, it looks like the `supportedManagers` entries are also affected by this bug:

- #13349

File: tools/docs/config.ts (hunk 4 above)
Comment (HonkingGoose) [reply tools/docs/config.ts:87]: Yeah, the new screenshot looks _way_ better. I can read and understand the content at once.

Syntax highlighting in the tables would be nice to have, but if you can't get it working, then we should still merge the prettier tables. 😉

Comment (Gabriel-Ladzaretti) [reply tools/docs/config.ts:87]: From what i gathered, html 5 doesn't support syntax highlight.
could we use something like [highlight.js](https://www.npmjs.com/package/highlight.js) perhaps?

What we saw in the markdown mocks is GH's extended markdown support for syntax highlighting

Comment (HonkingGoose) [reply tools/docs/config.ts:87]: Material for MkDocs supports syntax highlighting. I guess there might be a way to get it working in the table as well.

Here are the docs:

- https://squidfunk.github.io/mkdocs-material/setup/extensions/python-markdown-extensions/#highlight
- https://squidfunk.github.io/mkdocs-material/reference/code-blocks/

Comment (Gabriel-Ladzaretti) [reply tools/docs/config.ts:87]: When inspecting [packageRules](https://docs.renovatebot.com/configuration-options/#packagerules) for example.
The source of it is a ` ```json ...``` ` block inside the `configuration-options.md` file, while inside the generated mkdocs page we have this:
```html
<p><code>packageRules</code> is a powerful feature that lets you apply rules to individual packages or to groups of packages using regex pattern matching.</p>
<p>Here is an example if you want to group together all packages starting with <code>eslint</code> into a single branch/PR:</p>
<div class="highlight"><pre><span></span><code><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="nt">&quot;packageRules&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span><span class="w"></span>
<span class="w">    </span><span class="p">{</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;matchPackagePatterns&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span><span class="s2">&quot;^eslint&quot;</span><span class="p">],</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;groupName&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;eslint packages&quot;</span><span class="w"></span>
<span class="w">    </span><span class="p">}</span><span class="w"></span>
<span class="w">  </span><span class="p">]</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
</code></pre></div>
```

so its autogenerated while the docs are being constructed. 
Is there a way i can generated the mkdocs myself for experimenting?
it is worth checking if having this backticked block of code inside a table will produce the highlighted syntax we want.

Comment (viceice) [reply tools/docs/config.ts:87]: yes, checkout our docs repo und change the git submodule to your branch

File: tools/docs/config.ts (hunk 3 above)
Comment (Gabriel-Ladzaretti) [reply tools/docs/config.ts]: The outcome is a bit over engineered, as we need to support `${col}` and such.
Interesting topic nonetheless. 
we can revert it of course if needed.

File: tools/docs/config.ts [hunk 5]
```diff
@@ -1,11 +1,68 @@
...
 import { readFile, updateFile } from '../utils';
 
 const options = getOptions();
 
+type IndentTemplateFunction = (
+  strings: TemplateStringsArray,
+  ...keys: (string | number)[]
+) => string;
+
+function getIndentFunction(indent = '  '): IndentTemplateFunction {
```
[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply tools/docs/config.ts:14]: made it as generic as possible :), i don't mind hardcoding it and basically call `indent` directly

**Comment by Gabriel-Ladzaretti** [comment]
we now have syntax highlight 😀, see attached the mkdocs generated `index.html`
image
[index.zip](https://github.com/renovatebot/renovate/files/8894395/index.zip)

File: tools/docs/config.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (viceice) [reply tools/docs/config.ts:14]: yes, make it as simple as possible 🙃

**Review by viceice — APPROVED** [review]

File: tools/docs/config.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply tools/docs/config.ts]: made it simple but a tad bloated 😅

**Pull request merged**